swim/events.go

```diff
@@ -234,10 +229,48 @@ type RedundantReverseFullSyncEvent struct {
 	Target string `json:"target"`
 }
 
-func on(t interface{}, f func()) ListenerFunc {
+// DiscoHealEvent is sent when the discover provider healer attempts to heal a partition
+type DiscoHealEvent struct{}
+
+// AttemptHealEvent is sent when the healer is triggered
+type AttemptHealEvent struct{}
+
+// on is returns an EventListener that executes a function f upon receiving an
+// event of type t.
+func on(t interface{}, f func(e events.Event)) ListenerFunc {
 	return ListenerFunc(func(e events.Event) {
 		if reflect.TypeOf(t) == reflect.TypeOf(e) {
-			f()
+			f(e)
 		}
 	})
 }
+
+// ExecuteThenWaitFor executes a function and then waits for a specific type
+// of event to occur.
+//
+// Often we want to execute some code and then wait for an event be emitted due
+// to the code being executed. However in order to not miss the event we must
+// first register an event handler before we can execute the code:
+// - register listener that signals we can continue on receiving the correct event;
+// - execute the code that will lead to an event being emitted;
+// - wait for the continue signal.
+//
+// This can be quite hard to follow. Ideally we want it to look like
+// - execute the code that will lead to an event being emitted;
+// - and then wait for a specific event.
+//
+// This function helps with making the code read like the latter.
+func ExecuteThenWaitFor(f func(), er events.EventRegistrar, t interface{}) {
```

Comment:
I am still a bit scared by this code, it does not cause panic's anymore which is already better. However it causes a lot of garbage to accumulate in the event listener if this would be used outside of the tests. Since it is only used as a helper function during testing can we move it to a file ending in `_test.go` so it is available to tests, but can never be used by the library?